### DISCUSSION TITLE
Upgrade to Python 3.10 to use match statement.

### DISCUSSION THREAD

**Discussion by VoxelCubes** [ThioJoe/YT-Spammer-Purge#206] [Ideas]
Python finally added a very powerful switch statement called match in version 3.10, which released a few months ago. Are there some dependencies holding back on the version, or could this project upgrade to 3.10?
The match statement will simplify a lot of if,elif,else chains, of which there are many, so this would be the perfect application.
An example:
```Python
if filterChoice == "1" or filterChoice == "autosmart":
    filterMode = "AutoSmart"
elif filterChoice == "2" or filterChoice == "sensitivesmart":
    filterMode = "SensitiveSmart"
elif filterChoice == "3" or filterChoice == "id":
    filterMode = "ID"
elif filterChoice == "4" or filterChoice == "username":
    filterMode = "Username"
elif filterChoice == "5" or filterChoice == "text":
    filterMode = "Text"
elif filterChoice == "6" or filterChoice == "nameandtext":
    filterMode = "NameAndText"
elif filterChoice == "7" or filterChoice == "autoascii":
    filterMode = "AutoASCII"
```
becomes:
```Python
match filterChoice:
    case "1" | "autosmart":
        filterMode = "AutoSmart"
    case "2" | "sensitivesmart":
        filterMode = "SensitiveSmart"
    case "3" | "id":
        filterMode = "ID"
    case "4" | "username":
        filterMode = "Username"
    case "5" | "text":
        filterMode = "Text"
    case "6" | "nameandtext":
        filterMode = "NameAndText"
    case "7" | "autoascii":
        filterMode = "AutoASCII"
```

The end result are minor performance gains as well as greater ease reading and writing this code.

**Comment by somebadcode** [comment]
Good find! :+1: 
However, since you're doing that, you can with a default case reduce the amount of code even further.

```
diff --git a/YouTubeSpammerPurge.py b/YouTubeSpammerPurge.py
index 65d4370..0c98ecc 100644
--- a/YouTubeSpammerPurge.py
+++ b/YouTubeSpammerPurge.py
@@ -2342,28 +2342,26 @@ def main():
     else:
       filterChoice = input("\nChoice (1-7): ")
 
-    validChoices = ['1', '2', '3', '4', '5', '6', '7', 'id', 'username', 'text', 'nameandtext', 'autoascii', 'autosmart', 'sensitivesmart']
-    if filterChoice in validChoices:
-      validFilterMode = True
-      # Set string variable names for filtering modes
-      if filterChoice == "1" or filterChoice == "autosmart":
+    # Set string variable names for filtering modes
+    match filterChoice:
+      case "1" | "autosmart":
         filterMode = "AutoSmart"
-      elif filterChoice == "2" or filterChoice == "sensitivesmart":
-        filterMode = "SensitiveSmart"      
-      elif filterChoice == "3" or filterChoice == "id":
+      case "2" | "sensitivesmart":
+        filterMode = "SensitiveSmart"
+      case "3" | "id":
         filterMode = "ID"
-      elif filterChoice == "4" or filterChoice == "username":
+      case "4" | "username":
         filterMode = "Username"
-      elif filterChoice == "5" or filterChoice == "text":
+      case "5" | "text":
         filterMode = "Text"
-      elif filterChoice == "6" or filterChoice == "nameandtext":
+      case "6" | "nameandtext":
         filterMode = "NameAndText"
-      elif filterChoice == "7" or filterChoice == "autoascii":
+      case "7" | "autoascii":
         filterMode = "AutoASCII"
+      case _:
+        print(f"\nInvalid Filter Mode: {filterChoice} - Enter either 1, 2, 3, 4, 5, 6, or 7")
+        validConfigSetting = False
 
-    else:
-      print(f"\nInvalid Filter Mode: {filterChoice} - Enter either 1, 2, 3, 4, 5, 6, or 7")
-      validConfigSetting = False
 
   ## Get filter sub-mode to decide if searching characters or string
   validConfigSetting = None
```

However, the filters should be functions that return callables (`typing.Callable`) or just be callabe classes (implement `__call__`). That will decouple it more and the need for `match` disappears and there will be fewer conditionals and fewer places for spelling mistakes to shoot you in your foot. Increases clarity, extendibility and cohesion.

**Reply by VoxelCubes** [reply]
That would be ideal, using factories and protocols. This is a good example with a very similar use case demonstrated: https://youtu.be/s_4ZrtQs8Do

Only issue is that ThioJoe won't accept code he can't understand, so this tutorial will hopefully provide a good learning opportunity.

**Reply by somebadcode** [reply]
> That would be ideal, using factories and protocols. This is a good example with a very similar use case demonstrated: https://youtu.be/s_4ZrtQs8Do

Yeah, Arjan covers the newer features in a good way and it makes Python interesting to use.

> Only issue is that ThioJoe won't accept code he can't understand, so this tutorial will hopefully provide a good learning opportunity.

Sadly, yes. The current state of the script makes it possible to work with since the functions are too large, there are too many arguments and too little type hinting. Refactoring it using more Pythonic solutions is essentially a rewrite.
It's important to stick to the patterns that the language has because then more people can take part of the project and create their own extensions for it.

ThioJoe has also mentioned he's working on a rewrite in C# and that he will cross compile it for the various platforms. That raises some concerns since the people who wants to use this script want to be able to alter and extend it and Python is very beginner friendly. Cross compiling is excellent and C# in this case can show some nice performance improvements but I think it just goes too far. Cross compiling C# isn't very common, you rarely see .NET on Linux and Mac despite .NET 5.0 and .NET coming a long way on that front. So there are both good and bad things using C#.

**Comment by ThioJoe** [comment]
I actually had added a switch case with 3.10, but didn't end up using it more than like 2 places. So I reverted it in favor of more compatibility, since it didn't really save all that much space or add too much readability

**Reply by VoxelCubes** [reply]
Be that as it may, I would _highly_ recommend looking into the factory design pattern, as mentioned above, which will drastically simplify and make your code easier to maintain, when it comes to adding new processing steps. Instead of doing the whole filterMode = a string, then looking up what function to call with that string, you can assign a function to the filterMode and then just execute it at the end. Protocols help in that regard, which themselves are like a glorified type hint. Type hints in general help a lot with making code more maintainable, so you, the IDE, and anyone looking in, can immediately know what type something expects.

**Reply by VoxelCubes** [reply]
On a side note, python 3.10 also makes type hinting a bit easier, by letting you define unions with `int | str` instead of `from typing import Union` and then `Union[int, str]`. 
I'm not sure if all Linux distros use python 3.10 yet though, so maybe it's best to hold off after all.
